src/test/regress/sql/failure_multi_dml.sql

```diff
@@ -0,0 +1,148 @@
+COMMIT;
+
+--- shouldn't see any changes after failed DELETE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at INSERT
+SELECT citus.mitmproxy('conn.onQuery(query="^INSERT").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes before failed INSERT
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- fail at UPDATE
+SELECT citus.mitmproxy('conn.onQuery(query="^UPDATE").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- shouldn't see any changes after failed UPDATE
+SELECT * FROM dml_test ORDER BY id ASC;
+
+-- FIXME: Can't test this because the prepared transaction ID has PID in it
+
+-- fail at COMMIT (by failing PREPARE)
+-- SELECT citus.mitmproxy('conn.onQuery(query="^PREPARE").kill()');
+-- 
+-- BEGIN;
+-- DELETE FROM dml_test WHERE id = 1;
+-- DELETE FROM dml_test WHERE id = 2;
+-- INSERT INTO dml_test VALUES (5, 'Epsilon');
+-- UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+-- UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+-- COMMIT;
+-- 
+--- shouldn't see any changes after failed COMMIT
+-- SELECT * FROM dml_test ORDER BY id ASC;
+
+SELECT citus.mitmproxy('conn.allow()');
+
+-- drop table and recreate with different replication/sharding
+
+DROP TABLE dml_test;
+SET citus.shard_count = 1;
+SET citus.shard_replication_factor = 2; -- two placements
+
+CREATE TABLE dml_test (id integer, name text);
+SELECT create_distributed_table('dml_test', 'id');
+
+COPY dml_test FROM STDIN WITH CSV;
+1,Alpha
+2,Beta
+3,Gamma
+4,Delta
+\.
+
+---- test multiple statements against a single shard, but with two placements
+
+-- fail at COMMIT (actually COMMIT this time, as no 2pc in use)
+SELECT citus.mitmproxy('conn.onQuery(query="^COMMIT").kill()');
+
+BEGIN;
+DELETE FROM dml_test WHERE id = 1;
+DELETE FROM dml_test WHERE id = 2;
+INSERT INTO dml_test VALUES (5, 'Epsilon');
+UPDATE dml_test SET name = 'alpha' WHERE id = 1;
+UPDATE dml_test SET name = 'gamma' WHERE id = 3;
+COMMIT;
+
+--- should see all changes, but they only went to one placement (other is unhealthy)
+SELECT * FROM dml_test ORDER BY id ASC;
+SELECT shardid FROM pg_dist_shard_placement WHERE shardstate = 3;
+
+SELECT citus.mitmproxy('conn.allow()');
+
+-- drop table and recreate as reference table
+
+DROP TABLE dml_test;
+SET citus.shard_count = 2;
+SET citus.shard_replication_factor = 1;
+
+CREATE TABLE dml_test (id integer, name text);
+SELECT create_reference_table('dml_test');
+
+COPY dml_test FROM STDIN WITH CSV;
+1,Alpha
+2,Beta
+3,Gamma
+4,Delta
+\.
+
+-- fail at COMMIT (by failing to PREPARE)
+SELECT citus.mitmproxy('conn.onQuery(query="^PREPARE").kill()');
```

Comment:
Would it be worth to test the behaviour when a commit to a reference table fails? I guess it behaves like a distributed table, eg. recover after the 2pc is completed.

Reply:
Different test suite (there was a separate task for reference table modifications).